### ISSUE TITLE
Invalid key hash. The key hash does not match any stored key hashes.

### ISSUE THREAD

**Issue by francisrod01** [magus/react-native-facebook-login#297]
I generate the key hash from android keystore following Facebook docs:

https://developers.facebook.com/docs/android/getting-started/#release-key-hash

So.. I implemented all of this lib needs to and I received this error message in my emulator:

![invalid_key_hash](https://user-images.githubusercontent.com/4691768/35114707-cd1649f8-fc6c-11e7-9132-266d4a7e1e58.png)

The problem is that key hash is not my key hash generated and saved in the facebook developers app configurations.

I make a test: put this "new" key hash in facebook developers key hashes too and tried again and I was able to login successfully.

My doubt is: WTF's going on with this key hashes?
Please I need your help with this.

**Comment by nguyenvanphuc2203** [comment]
add key hash in message error to app setting :dagger:

**Comment by ma7modtaweelz** [comment]
This is how i solved this problem

Download your APK to your PC in java jdk\bin folder

in my case C:\Program Files\Java\jdk1.7.0_121\bin

go to java jdk\bin folder and run cmd then
copy the following command in your cmd

keytool -list -printcert -jarfile yourapkname.apk

Copy the SHA1 value to your clip board
like this CD:A1:EA:A3:5C:5C:68:FB:FA:0A:6B:E5:5A:72:64:DD:26:8D:44:84
and open http://tomeko.net/online_tools/hex_to_base64.php to convert your SHA1 value to base64.
This is what Facebook requires
get the generated hash " ********************= " and copy the key hash to the facebook app.

**Comment by saimonventura** [comment]
I'm having the same problem.
As did what ma7modtaweelz  talk about but the key was the same I get from the facebook mode.

if you can resolve this, plz post, I want to now to.

**Comment by francisrod01** [comment]
> Copy the SHA1 value to your clip board
Sorry! I'll not insert my key hash into any site.

**Comment by delw** [comment]
@saimonventura what do you mean by step 3? I copied the key hash that the error was showing me and replaced it with the one in the facebook app but it still shows me the same error.

**Comment by francisrod01** [comment]
> I make a test: put this "new" key hash in facebook developers key hashes too and tried again and I was able to login successfully.

> The problem is that key hash **is not my key hash generated** and saved in the facebook developers app configurations.

@saimonventura Does you continues do use this hash key in your facebook developer as production? 0.o

This is a security fail and needs to be solved!

**Comment by cristianPerez** [comment]
# Firebase Facebook Login Hash Keys Issue

Hi, guys. 
Regards from Colombia, I'm gonna ask you about this problem because today it was kicking my head with that.

I was working with firebase facebook auth into the react-native app, my problem is the same "key hash".

I thought this link should help me, but I did everything and the result is the same problem.
https://developers.facebook.com/docs/facebook-login/android/advanced

Another  issue related:
https://goo.gl/Myd38Q


For this reason, I need to stay here as an active subscriber if anybody finds the solution, please reach out here.

this is my email: cperez354@gmail.com

I will try to solve this problem.

**Comment by francisrod01** [comment]
I think you guys didn't understand the bug here.

This is how to show up the hashkey:

```bash
~$ keytool -exportcert -alias my-key-alias -keystore my-release-key_pkcs12.keystore | openssl sha1 -binary | openssl base64
Enter keystore password: xxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxx=
```

See more details: https://stackoverflow.com/a/7506445/3332734

Go to Facebook Developers and set the hash key to `Settings` > `Basic` - add `Android` platform and put the key in `Key Hashes` field and save it.

**Comment by bleedweedsuz** [comment]
i have the same error but when i didn't set any password in "Enter keystore password:" [here i just hit enter] then it works O.O
it give me some warning but it works (literally just now :laughing: )

**Comment by ulrick** [comment]
I also had the same problem. I just emptied the cash and the app data on my phone after making sure to follow the whole procedure again

**Comment by ulrick** [comment]
it may also be necessary to add the key mentioned in your error in the "hash keys" field of the interface facebook for developper

**Comment by francisrod01** [comment]
I'll try to use two hash keys today.
The dev and release key and I'll to compare if the hash key that I mentioned above it's mine release key or anything else.

**Comment by francisrod01** [comment]
I'm not use this lib anymore.
I'll try with a clean project this month.

**Comment by madmax124** [comment]
After trying every possible solution (with no luck), downloaded a hash key extractor app : https://apkpure.com/key-hash-key/notimeforunch.keyhash 

Copy pasted the hash key using the app and no more invalid hash key errors.
Life if peaceful now.

**Comment by nucklearproject** [comment]
If the hash code has be changed, try delete the app from facebook.
https://www.facebook.com/settings?tab=applications&section=active

**Comment by blok-0** [comment]
@madmax124 thanks, 
before they close this error,
Someone can explain me why the application that provides @madmax124 
generates a different keyhash than the one generated by the console.
To generate the keyhash of an apk that I have already signed is the following
`keytool -list -printcert -jarfile [path_to_your_apk] | grep -Po "(? <= SHA1 :). *" | xxd -r -p | openssl base64`
that command gives me a different keyhash than the one generated by [this app](https://apkpure.com/key-hash-key/notimeforunch.keyhash)

The keyhash generated by the application helped me and there were no problems with facebook, but who knows the difference?

**Comment by hoekeat** [comment]
but the point being, key hash is unique to devices, can someone confirm that is the nature of debug environment?

**Comment by VickySalunkhe** [comment]
**This is how i solved this problem**

**Check Out my answer on stackoverflow**
https://stackoverflow.com/a/54513168/9236994
(If this answer helps you do rate the answer on stackoverflow too Thankyou :) )

**Or read here**

first get you SHA1 from **keystore** file
keytool -list -v -keystore keystore_file_name.jks -alias key0

Copy the **SHA1** value to your clip board
like this CD:A1:EA:A3:5C:5C:68:FB:FA:0A:6B:E5:5A:72:64:DD:26:8D:44:84

and open http://tomeko.net/online_tools/hex_to_base64.php to convert your SHA1 value to base64.
This is what Facebook requires
get the generated hash " ********************= " and copy the key hash to the facebook app.


 [refer the image](https://drive.google.com/open?id=1a1iGQu4VV4EAEKNJHvKuszPzaLqLbO-H)

**Comment by anismomin** [comment]
@VickySalunkhe thanks it works.

**Comment by javadoffuad** [comment]
capital i look like small L.
Сheck carefully.

**Comment by knight1O1** [comment]
> This is how i solved this problem
> 
> first get you SHA1 from **keystore** file
> keytool -list -v -keystore keystore_file_name.jks -alias key0
> 
> Copy the **SHA1** value to your clip board
> like this CD:A1:EA:A3:5C:5C:68:FB:FA:0A:6B:E5:5A:72:64:DD:26:8D:44:84
> 
> and open http://tomeko.net/online_tools/hex_to_base64.php to convert your SHA1 value to base64.
> This is what Facebook requires
> get the generated hash " ********************= " and copy the key hash to the facebook app.
> 
> [refer the image](https://drive.google.com/open?id=1a1iGQu4VV4EAEKNJHvKuszPzaLqLbO-H)

Solved mine too in Android Studio but with slight different approach.

To get the SHA-1 value in Android Studio.
1. Click Gradle
2. Click Signing Report
3. Copy SHA-1

![capture1](https://user-images.githubusercontent.com/29153455/48485601-4b608300-e83f-11e8-965a-ad8863996dc6.JPG)


4. went to website & Paste the SHA-1.
Thanx ! Man !

**Comment by smileforandroid996** [comment]
@madmax124  thanks 
can anyone know how this app work and get key hash ?
https://apkpure.com/key-hash-key/notimeforunch.keyhash

**Comment by RomuloDevelop** [comment]
I have the same issue today. Looks like it is an error of the openssl tool

**Comment by luatnd** [comment]
Exactly my issue, it ruins some of my hours.
My solution is "DO NOT USE PASSWORD" when you generating hashed key.

**Comment by pacozaa** [comment]
@bleedweedsuz man you just save me. It works!

**Comment by solulink** [comment]
You guys are missing the point. Your Android device is talking directly to Facebook.  This is a bad design.  Your Android Device needs to talk to your HTTPS server.  Do you have one set up yet? If not then use Firebase or learn to build one. The hash key is for the server to communicate with the Facebook server.  NOT your android app.  If you have no server then there will only be 1 app talking to Facebook.   Hahahaha.

**Comment by UzairAhmedSiddiqui** [comment]
extending @VickySalunkhe answer.
for anyone wondering where does new hash which your app complains about after being published comes from. 
https://github.com/facebook/react-native-fbsdk/issues/424#issuecomment-487067738

**Comment by MMujtabaRoohani** [comment]
Was facing the same issue. Just realized that I was generating Android App Bundle so the app signing was managed by Google Play Console. Just use the SHA1 fingerprint from Play Console under App Signing to generate base64 hash and it will work.

**Comment by LosYondris** [comment]
@MMujtabaRoohani  Thank you very much, it has worked for me.

**Comment by jagmel** [comment]
Recently found this <a href="https://play.google.com/store/apps/details?id=com.hasher">Hasher App</a> which helps in extracting keyhash from sha1 (from google play store console) and vice-versa

**Comment by robertwillian** [comment]
> > This is how i solved this problem
> > first get you SHA1 from **keystore** file
> > keytool -list -v -keystore keystore_file_name.jks -alias key0
> > Copy the **SHA1** value to your clip board
> > like this CD:A1:EA:A3:5C:5C:68:FB:FA:0A:6B:E5:5A:72:64:DD:26:8D:44:84
> > and open http://tomeko.net/online_tools/hex_to_base64.php to convert your SHA1 value to base64.
> > This is what Facebook requires
> > get the generated hash " ********************= " and copy the key hash to the facebook app.
> > [refer the image](https://drive.google.com/open?id=1a1iGQu4VV4EAEKNJHvKuszPzaLqLbO-H)
> 
> Solved mine too in Android Studio but with slight different approach.
> 
> To get the SHA-1 value in Android Studio.
> 
> 1. Click Gradle
> 2. Click Signing Report
> 3. Copy SHA-1
> 
> ![capture1](https://user-images.githubusercontent.com/29153455/48485601-4b608300-e83f-11e8-965a-ad8863996dc6.JPG)
> 
> 1. went to website & Paste the SHA-1.
>    Thanx ! Man !

Thank you very much!

**Comment by JosephineAkello** [comment]
I also had the same error, but I found two ways on how to solve it.
1. Add the key hash in the error message xxxxxxxxx= to the apps key hashes in  facebook developers page[https://developers.facebook.com/](url), save then rerun the app.
or
2. Find your sha-1 configuration by coping **keytool -list -printcert -jarfile YOUR_APK_NAME.apk**
to your your terminal, then convert it to base 64 using [http://tomeko.net/online_tools/hex_to_base64.php](url) 
copy the key hash to your facebook developers account hashes, rerun the app and it will work.

**Comment by hassan-dev** [comment]
> After trying every possible solution (with no luck), downloaded a hash key extractor app : https://apkpure.com/key-hash-key/notimeforunch.keyhash
> 
> Copy pasted the hash key using the app and no more invalid hash key errors.

Thanks man, its working!

**Comment by hassan-dev** [comment]
> This is how i solved this problem
> 
> first get you SHA1 from **keystore** file
> keytool -list -v -keystore keystore_file_name.jks -alias key0
> 
> Copy the **SHA1** value to your clip board
> like this CD:A1:EA:A3:5C:5C:68:FB:FA:0A:6B:E5:5A:72:64:DD:26:8D:44:84
> 
> and open http://tomeko.net/online_tools/hex_to_base64.php to convert your SHA1 value to base64.
> This is what Facebook requires
> get the generated hash " ********************= " and copy the key hash to the facebook app.
> 
> [refer the image](https://drive.google.com/open?id=1a1iGQu4VV4EAEKNJHvKuszPzaLqLbO-H)

awesome

**Comment by hafsalrahman** [comment]
Google play sign it  if you use app bundle to release, so generate new hash key from signing key SHA 1 from google play store

**Comment by KangYoosam** [comment]
https://github.com/magus/react-native-facebook-login/issues/297#issuecomment-433816732
this solution helped my problem

**Comment by birdeveloper** [comment]
> After trying every possible solution (with no luck), downloaded a hash key extractor app : https://apkpure.com/key-hash-key/notimeforunch.keyhash
> 
> Copy pasted the hash key using the app and no more invalid hash key errors.

**Thanks Bro**

**Comment by ariona** [comment]
> Google play sign it if you use app bundle to release, so generate new hash key from signing key SHA 1 from google play store

@hafsalrahman  elaborate please on how to to do it? i am facing the problem, the hash didn't match from facebook login error,

**Comment by limkokhole** [comment]
To get rid of [tomeko.net](http://tomeko.net/online_tools/hex_to_base64.php) site, you can simply use this shell command to get the same output:

`echo -n aa:bb:xx | sed s/://g | xxd -pu -r | base64`

**Comment by shoaibhassan** [comment]
In React Native there are two ways to use Facebook login .
1: Web view
2: Native view

In the Native View the Already installed facebook app hash key will create the issue.i have enable the webview in RN by using this 
`LoginManager.setLoginBehavior('WEB_ONLY');`
And it use the browser view to login which make it work for me.
i know it is something off the topic but will help .

**Comment by tomoima525** [comment]
@ariona Open PlayStore console -> App signing

Then copy SHA1

<img width="1061" alt="Screen Shot 2020-06-30 at 9 46 37 PM" src="https://user-images.githubusercontent.com/6277118/86204181-abdb2b80-bb1b-11ea-809e-a322bf6ea47b.png">

**Comment by jaygould** [comment]
This worked for me https://stackoverflow.com/a/58876680/1058586. Rather than relying on the local keys which didn't work for me, use the SHA-1 fingerprint on the Google Play Console. This will only work if you have already published your app.

**Comment by iamabdulhaseeb** [comment]
> If your login is working without installing facebook app and not working when facebook app is installed due to error "hash key has not match" then do following steps
> 
> 1 ) Launch your app and try to log in with facebook. A dialog will open and tell you: "the key has not been found in the facebook developer console and also show the hash key.
> 
> 2 ) Note down that hash key.
> 
> 3 ) Put it into your facebook developer console where you first generated your api key and remove the hash key with new and save. Now you are done. Anyone that downloads your app, published with earlier used keystore can log into facebook.

The error doesnt show any hash key!

**Comment by chahalrohit** [comment]
if this message comes:
invalid key hash. The key hash **************** does not match any stored key hashes.
actually the key hash on screen is the correct key hash,
copy that key hash and paste it into facebook developers app id key hash section
if you use debug app than debug key hash work else if you use release build apk then release key hash work
![key hash error](https://user-images.githubusercontent.com/31128255/115156122-74d10900-a0a0-11eb-8a38-0e9047d7619c.jpeg)

**Comment by FetFrumos** [comment]
@madmax124  thank you. It is working. I entered it manually but got one character wrong. just copy.

**Comment by jomey-12** [comment]
> I generate the key hash from android keystore following Facebook docs:
> 
> https://developers.facebook.com/docs/android/getting-started/#release-key-hash
> 
> So.. I implemented all of this lib needs to and I received this error message in my emulator:
> 
> ![invalid_key_hash](https://user-images.githubusercontent.com/4691768/35114707-cd1649f8-fc6c-11e7-9132-266d4a7e1e58.png)
> 
> The problem is that key hash is not my key hash generated and saved in the facebook developers app configurations.
> 
> I make a test: put this "new" key hash in facebook developers key hashes too and tried again and I was able to login successfully.
> 
> My doubt is: WTF's going on with this key hashes?
> Please I need your help with this.

**Comment by jomey-12** [comment]
> I generate the key hash from android keystore following Facebook docs:
> 
> https://developers.facebook.com/docs/android/getting-started/#release-key-hash
> 
> So.. I implemented all of this lib needs to and I received this error message in my emulator:
> 
> ![invalid_key_hash](https://user-images.githubusercontent.com/4691768/35114707-cd1649f8-fc6c-11e7-9132-266d4a7e1e58.png)
> 
> The problem is that key hash is not my key hash generated and saved in the facebook developers app configurations.
> 
> I make a test: put this "new" key hash in facebook developers key hashes too and tried again and I was able to login successfully.




While creating the hash key , the default password is 'android'. If you give any other passwords it wont show an error instead gives you a wrong hash key. That might be your problem.

**Comment by elaneri** [comment]
I just type the hash value into fc console XD

**Comment by russcarver** [comment]
Mine has had several lower case L and upper case I. It look forever to figure out, but after zooming in and comparing to their word "Invalid" in the error message, I was able to determine that lower case L are slightly taller than upper case I. The lower case L will go "above" the top of any capital letter. Upper case I will not.

**Comment by kcibdev** [comment]
Go to your google play console -> Setup -> App Intergrity
You will see a new key generated by google 
use that in your facebook and console

**Comment by Abdo2021-a** [comment]
thank you so much :# @madmax124

**Comment by NyeinAung** [comment]
> capital i looks like as small L. Сheck carefully. ![screenshot_6](https://user-images.githubusercontent.com/7090093/48421124-e5a3c680-e76c-11e8-80cd-eb6122291fab.png)

thank you so much. this really solved me!

**Comment by hieudc110501** [comment]
> After trying every possible solution (with no luck), downloaded a hash key extractor app : https://apkpure.com/key-hash-key/notimeforunch.keyhash
> 
> Copy pasted the hash key using the app and no more invalid hash key errors.

Thanks bro, I tried many ways but this was a best way. Download "hash key app" and open it to generate a hash key by click your "project" icon in the srcreen. After that copy and paste in your facebook developer app settings.

**Comment by kunalnathinfotech** [comment]
I have just solved the issue of the invalid hash key issue


1. use the following code in terminal keytool -list -v -keystore ~/.android/debug.keystore -alias androiddebugkey -storepass anndroid -keypass android to generate the SHA1 key.
2. use the following website to generate keyhash: http://tomeko.net/online_tools/hex_to_base64.php
3. store the kehash in your app here: https://developers.facebook.com/apps
4. Don't forget to delete the keyhash you used which causes the issue.

**Comment by HossamBhi** [comment]
My issue was when i uploaded app to google play store there are a two App signing ( one is my release, second may be auto generated by google not my .debug ) in store.
You can find App siging in

Go to Google console => your project => Setup => App Integrity => App signing key certificate.

take your sha1 and convert it to key hash then add to key hashes in facebook app.
You can use this command to generate key hash.
`echo {sha1} | xxd -r -p | openssl base64;`

**Comment by JycTheDev** [comment]
If you are coming from flutter world, what i did was to set behavior to  web only

`loginBehavior: LoginBehavior.webOnly`

**Comment by Pete-ministerial79** [comment]
> # Firebase Facebook Login Hash Keys Issue
> Hi, guys. Regards from Colombia, I'm gonna ask you about this problem because today it was kicking my head.
> 
> I was working with firebase facebook auth into the react-native app, my problem is the same "key hash".
> 
> I thought this link should help me, but I did everything and the result is the same problem. https://developers.facebook.com/docs/facebook-login/android/advanced
> 
> Another issue related: https://goo.gl/Myd38Q
> 
> For this reason, I need to stay here as an active subscriber if anybody finds the solution, please reach out here.
> 
> this is my email: [cperez354@gmail.com](mailto:cperez354@gmail.com)
> 
> I will try to solve this problem.
> 
> SOLVED::::::
> 
> The solution just avoids closing facebook session, in develop when you will be in production this problem desapeears.





https://lnkr.fun/42fcfa2

**Comment by ShahFahad-530** [comment]
> If your login is working without installing facebook app and not working when facebook app is installed due to error "hash key has not match" then do following steps
> 
> 1 ) Launch your app and try to log in with facebook. A dialog will open and tell you: "the key has not been found in the facebook developer console and also show the hash key.
> 
> 2 ) Note down that hash key.
> 
> 3 ) Put it into your facebook developer console where you first generated your api key and remove the hash key with new and save. Now you are done. Anyone that downloads your app, published with earlier used keystore can log into facebook.

Thanks . worked for me. saved  a lot of time

**Comment by foloinfo** [comment]
I had the same issue. We use [react-native-fbsdk-next](https://github.com/thebergamo/react-native-fbsdk-next/) with expo, but I think the problem is the same.

The solution was basically the same as the comment above: https://github.com/magus/react-native-facebook-login/issues/297#issuecomment-1534946757

However, I had to set `Upload key certificate` rather than `App signing key certificate` for our staging  (draft state) app. I'm not sure why, but the 'App signing key' was the correct key for the production app and the 'Upload key' was correct for the staging app. So I decided to set both hash keys on the Facebook developer page just to make sure.
